### DISCUSSION TITLE
Executing files from flash

### DISCUSSION THREAD

**Discussion by paulober** [micropython/micropython#15760] [General]
I wanted to ask what is the recommended option to run files form the filesystem?
I found those three options:
- `execfile(fn)` is deprecated in normal Python and requires a compiler flag
- `exec(open(...).read())` recommended alternative to first option in normal Python but seems like it would load the full file into memory
- `import myfile` does not run if __name__ == "__main__" and requires more work to run a second time as you first have to remove the module

**Comment by peterhinch** [comment]
If the file is a .py script, it will be loaded into memory so that the compiler can compile it to bytecode. This applies whatever method you use. The RAM used by the source will then be reclaimed with the bytecode being run from RAM. The compilation phase may be skipped by precompiling, in which case the bytecode will be loaded directly into RAM. To run from flash with near zero RAM usage you need frozen bytecode.

It is possible to reload a module (but maybe you already know this):
```py
import gc
import sys
def reload(mod):
    mod_name = mod.__name__
    del sys.modules[mod_name]
    gc.collect()
    __import__(mod_name)
```

**Reply by paulober** [reply]
Thanks for the response. It's good to know that `execfile` and `import` also load the full content into memory.

**Reply by Josverl** [reply]
@Paul, Ive used `exec(open(...).read())` in the past in some rather primitive IDEs to add an option execute a specific file from a file tree.
I assume you might be thinking about something similar.

I found I works well - but memory fragmentation can be a challenge as these files get larger, especially on systems with less memory.   
Also I think that the users that use such functionality, tend to use a `single file` development approach rather than a `module based`. 
So definitely useful, but also something to educate your users on how to use , and when to split off code to a module and import it.
